File: leetcode/array/46_permutations.go
package array

// type stack []int

// func (s stack) Push(v int) stack {
//     return append(s, v)
// }

// func (s stack) Pop() (stack, int) {
//     l := len(s)
//     return  s[:l-1], s[l-1]
// }




func Permute(nums []int) [][]int {
	var res [][]int
	res = dfs(nums, res)
	return res
}


func dfs(nums []int, res [][]int) [][]int{
	if len(nums) == 1 {
		i := [][]int{nums[:]}
		return i
	}

	for i:= 0; i < len(nums); i++ {
		n := nums[0]
		nums = nums[1:]
		perms := dfs(nums, res)
		var temp [][]int
		for _,perm := range perms {
			perm = append(perm, n)
			temp = append(temp, perm)	
		res = append(res, temp...)
		nums = append(nums, n)	
		}
	}
	return res
}

File: leetcode/array/46_permutations.py

def permute(nums):
    result = []
    # base case
    if(len(nums)==1):
        return [nums[:]]

    for i in range(len(nums)):
        n = nums.pop(0)        # 1
        perms = permute(nums)  # [[2,3], [3,2]]  
        for perm in perms:
            perm.append(n)     # [[2,3,1], [3,2,1]]  
        result.extend(perms)
        nums.append(n)

    return result


nums = [1,2,3]
print(permute(nums))